### PR TITLE
Fix compatibility of examples with py3

### PR DESCRIPTION
**Pull request by rahul003**
## Description ##
Fixes compatibility issues of example/rcnn and example/sparse and example/gluon/tree_lstm to run with python3

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
To ensure compatibility of example/rcnn with py2.7 and py3.6
- [x] Change cPickle to pickle for py3
- [x] Remove xrange to use builtins.range
- [x] Change how modules are imported
- [x] Update usage of numpy to ensure we are not calling deprecated APIs. Example using float to index an array has been removed from numpy. Now we need to cast it to int first.
- [x] Updated readme

To ensure compatibility of example/sparse with py3.6
- [x] An assert statement was comparing string and float causing type error on py3.6

To ensure example/sparse runs on Mac
- [x] Updates scripts to fetch data to use common API for Mac and Linux. Removed wget.
- [x] Updated get_bz2_data to extract file line by line. Else it fails on Mac as MacOS doesn't support that feature for large files.
- [x] Added get_zip_data in test_utils, which works on both Mac and Linux

## Comments ##
- example/rcnn now needs pip packages builtins even for py2.7 so we can use same API for both py2.7 and py3.6. This is not a big deal since this is an advanced example and already requires 4-5 pip packages to run

@anirudh2290 especially for test_utils changes
@eric-haibin-lin especially for sparse changes #

### CODE REVIEW

**Comment by marcoabreu** [comment]
Hello @rahul003, please make sure to rebase your PR.

**Comment by rahul003** [comment]
Rebased

**Comment by rahul003** [comment]
@marcoabreu what instance types is the CI using? We seem to be running out of memory. The error in CI is for test_sgd running out of cuda memory. This code has nothing to do with that error.

**Comment by marcoabreu** [comment]
@rahul003 We're using G3.8xlarge with 8Gb of CUDA-memory - that should be more than sufficient. I've submitted a PR to disable this test temporarily: https://github.com/apache/incubator-mxnet/pull/9001
In the meantime, I retriggered your build. This was a flaky test.

File: example/rcnn/rcnn/core/tester.py [hunk 1]
```diff
@@ -15,13 +15,14 @@
 # specific language governing permissions and limitations
 # under the License.
 
-import cPickle
+from six.moves import cPickle
```
Comment (piiswrong) [inline example/rcnn/rcnn/core/tester.py]: can you do it without adding six?

Comment (rahul003) [reply example/rcnn/rcnn/core/tester.py]: this example already requires few libraries to be installed. Installing six doesn't seem too hard? Besides not using six means using ugly if else statements for different versions of python

**Comment by rahul003** [comment]
Done

**Pull request merged**